### PR TITLE
DDF-05147: Fixed null date handling in `DateLongFormatTypeAdapter`

### PR DESCRIPTION
**Pull request by leo-sakh**
#### What does this PR do?
Fixed GSON type adapter null handling

#### Who is reviewing it? 
@bdeining 
@jrnorth 
@troymohl 
@rymach 

#### Select relevant component teams: 
@codice/io 

#### Ask 2 committers to review/merge the PR and tag them here.
@bdeining
@jrnorth

#### How should this be tested?
1. Create search and/or result form
2. Perform search using the form from 1.
3. Observe no NPE in karaf log output
4. Valid search results returned

#### What are the relevant tickets?
Fixes: #5147 

#### Checklist:
- [ ] Documentation Updated
- [ ] Update / Add Threat Dragon models
- [ ] Update / Add Unit Tests
- [ ] Update / Add Integration Tests

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.

### CODE REVIEW

File: libs/gson-support/src/main/java/org/codice/gsonsupport/GsonTypeAdapters.java [hunk 1]
```diff
@@ -46,7 +47,8 @@ private GsonTypeAdapters() {}
   public static class DateLongFormatTypeAdapter extends TypeAdapter<Date> {
     @Override
     public void write(JsonWriter out, Date date) throws IOException {
-      out.value(date.getTime());
+      Long time = Optional.ofNullable(date).map(Date::getTime).orElse(null);
```
Comment (bdeining) [inline libs/gson-support/src/main/java/org/codice/gsonsupport/GsonTypeAdapters.java]: Is there a unit test we can add for this?

Comment (leo-sakh) [reply libs/gson-support/src/main/java/org/codice/gsonsupport/GsonTypeAdapters.java]: Added some basic ones around this issue.

File: libs/gson-support/src/test/java/org/codice/gsonsupport/GsonTypeAdaptersTest.java [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            .registerTypeAdapter(Date.class, new GsonTypeAdapters.DateLongFormatTypeAdapter())
+            .create();
+  }
+
+  @Test
+  public void testReadingValidLongDate() {
+    String jsonString = "{\"currentDate\": 1546344000000}";
+    GsonTypeAdaptersTestPojo parsedDate = gson.fromJson(jsonString, GsonTypeAdaptersTestPojo.class);
+
+    assertThat(parsedDate.currentDate.toString(), equalTo("Tue Jan 01 07:00:00 EST 2019"));
```
Comment (jrnorth) [inline libs/gson-support/src/test/java/org/codice/gsonsupport/GsonTypeAdaptersTest.java]: Is this dependent on the computer's time zone?

Comment (leo-sakh) [reply libs/gson-support/src/test/java/org/codice/gsonsupport/GsonTypeAdaptersTest.java]: Good catch! Fixed

File: libs/gson-support/src/test/java/org/codice/gsonsupport/GsonTypeAdaptersTest.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  }
+
+  @Test
+  public void testReadingValidLongDate() {
+    String jsonString = "{\"currentDate\": 1546326000000}";
+    GsonTypeAdaptersTestPojo parsedDate = gson.fromJson(jsonString, GsonTypeAdaptersTestPojo.class);
+
+    Instant parsedInstant = parsedDate.currentDate.toInstant().atOffset(ZoneOffset.UTC).toInstant();
+
+    assertThat(parsedInstant.toEpochMilli() + "", equalTo("1546326000000"));
```
Comment (jrnorth) [inline libs/gson-support/src/test/java/org/codice/gsonsupport/GsonTypeAdaptersTest.java]: `assertThat(parsedInstant.toEpochMilli(), equalTo(1546326000000L))`?

**Comment by jrnorth** [comment]
build now

**Comment by leo-sakh** [comment]
build now

File: libs/gson-support/pom.xml [hunk 4]
```diff
@@ -43,4 +43,45 @@
...
+                        <configuration>
+                            <haltOnFailure>true</haltOnFailure>
+                            <rules>
+                                <rule>
+                                    <element>BUNDLE</element>
+                                    <limits>
+                                        <limit implementation="org.codice.jacoco.LenientLimit">
+                                            <counter>INSTRUCTION</counter>
+                                            <value>COVEREDRATIO</value>
+                                            <minimum>0.01</minimum>
```
Comment (troymohl) [inline libs/gson-support/pom.xml]: Should we just leave this 0.0 like the others? Also, is the coverage really this low on this module?

Comment (leo-sakh) [reply libs/gson-support/pom.xml]: There's no coverage on this module outside of what I added.

**Comment by jrnorth** [comment]
build now

**Comment by jrnorth** [comment]
build now

**Comment by jrnorth** [comment]
`download-maven-plugin` failed. build now

**Comment by leo-sakh** [comment]
build now

**Comment by jrnorth** [comment]
Looks like some kind of Docker issue. build now

**Comment by jrnorth** [comment]
Network failure. build now

**Comment by gjvera** [comment]
build now

**Comment by gjvera** [comment]
### 🎉 Hero Successful 🎉 
- created search forms/result forms
- was able to use them successfully in intrigue 

LGTM!

**Comment by jrnorth** [comment]
The build only failed because Solr didn't stop correctly.
```
18:05:56 [INFO] ------------------------------------------------------------------------
18:05:56 [INFO] BUILD SUCCESS
18:05:56 [INFO] ------------------------------------------------------------------------
18:05:56 [INFO] Total time: 05:57 h
18:05:56 [INFO] Finished at: 2019-08-17T01:05:56+00:00
18:05:56 New run name is 'PR 5149'
18:05:56 Taking single-use slave mesos-jenkins-524e91e65f7e4de2b96eae5d7c2d3d58-linux-large offline.
18:05:56 Build was aborted
18:05:59 [INFO] Final Memory: 592M/1587M
18:05:59 [INFO] ------------------------------------------------------------------------
18:05:59 [ERROR] Async process failed for: [/jenkins/workspace/DDF-Jobs/pr/Linux/distribution/test/itests/test-itests-ddf/target/solr/bin/solr, stop, -p, 9784]
18:05:59 org.apache.commons.exec.ExecuteException: Process exited with an error: 143 (Exit value: 143)
18:05:59 	at org.apache.commons.exec.DefaultExecutor.executeInternal(DefaultExecutor.java:404)
18:05:59 	at org.apache.commons.exec.DefaultExecutor.access$200(DefaultExecutor.java:48)
18:05:59 	at org.apache.commons.exec.DefaultExecutor$1.run(DefaultExecutor.java:200)
18:05:59 	at java.lang.Thread.run(Thread.java:748)
```

**Pull request merged**